packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts

```diff
@@ -459,12 +361,258 @@ export class OdspDocumentStorageService implements IDocumentStorageService {
         });
     }
 
-    private async fetchSnapshot(snapshotOptions: string, tokenFetchOptions: TokenFetchOptions) {
+    private isTreesLatestSecondCallEvent(tokenFetchOptions: TokenFetchOptions) {
+        if (tokenFetchOptions.refresh) {
+            // This is the most critical code path for boot.
+            // If we get incorrect / expired token first time, that adds up to latency of boot
+            this.logger.sendErrorEvent({ eventName: "TreeLatest_SecondCall", hasClaims: !!tokenFetchOptions.claims });
+        }
+    }
+
+    private async fetchSnapshotViaPost(
+        tokenFetchOptions: TokenFetchOptions,
+        hostPolicy: ISnapshotOptions,
+    ): Promise<IOdspSnapshot> {
+        let cachedSnapshot: IOdspSnapshot | undefined;
+
+        // No need to ask cache twice - if first request was unsuccessful, cache unlikely to have data on second turn.
+        if (tokenFetchOptions.refresh) {
+            cachedSnapshot = await this.fetchSnapshotViaPostCore(hostPolicy, tokenFetchOptions);
+        } else {
+            cachedSnapshot = await PerformanceEvent.timedExecAsync(
+                this.logger,
+                { eventName: "ObtainSnapshot" },
+                async (event: PerformanceEvent) => {
+                    const cachedSnapshotP = this.cache.persistedCache.get(
```

Comment:
You "forked" code based on POST vs. GET too early, which results in two instanced of code that deals with cache layer.
I believe prior workflow was better and we should revert, i.e. POST vs. GET should be encapsulated in fetchSnapshot() API - I do not see the reason to make that fork that early, and in getSnapshotTree()

Reply:
Changed the logic.